Added PreviewLinkAutopopulate plugin to auto-populate preview links

stephen-cox opened the pull request:
Part of localgovdrupal/localgov#600

stephen-cox committed:
Added PreviewLinkAutopopulate plugin to auto-populate preview links localgovdrupal/localgov#600

stephen-cox committed:
Fixed typo

stephen-cox committed:
Merge branch '2.x' into feature/2.x/localgov-600-autopopulate-preview-links

ekes commented:
Is there something else that needs to be done other than checking out the branch. I've got it, and done a site install, and used demo, but I don't see the button, only one to copy the preview link.

Also is it intentionally only getting the next level of child pages, or should it be getting all child pages (using the Nested Storage)?

stephen-cox commented:
> Also is it intentionally only getting the next level of child pages, or should it be getting all child pages (using the Nested Storage)?

I missed that subsite pages can also be parents along with overviews, I'll have to update to logic to handle this.

finnlewis commented:
@stephen-cox working on this today

stephen-cox committed:
Added new method to fetch flatterned entity hierachy to SubsitesHierarchyTrait

stephen-cox committed:
Added test for new getFlattenedSubsiteHierarchy method

stephen-cox commented:
@ekes I have updated this PR to use Entity Hierarchy to load the nodes in a subsite; which is frankly a pain to work with. Loading all nodes is not straightforward so I've added a test.

Would be nice to get this reviewed before Next Tuesday, if possible, as it's been hanging around for a while now.

ekes commented on src/Plugin/Block/SubsitesHierarchyTrait.php:
```diff
@@ -123,4 +123,41 @@ protected function getRootId(EntityInterface $entity): ?int {
+
+    $storage = $this->getNestedSetStorage('localgov_subsites');
+    $node_key = $this->getNestedSetNodeKeyFactory()->fromEntity($node);
+    if ($ancestors = $storage->findAncestors($node_key)) {
```
This could just be $storage->findRoot($node_key) which does much the same.

ekes reviewed (approved):
For the rest that works.

ekes commented:
> Loading all nodes is not straightforward so I've added a test.

It's just get root, get all ancestors of root, and load them? Seems straightforward to me. Pain I found is the way that NestedSet class has be abstracted so you don't get code completion (or I don't anyway).
